lib/foreman_remote_execution_core/script_runner.rb

```diff
@@ -215,29 +223,33 @@ def ensure_local_directory(path)
     end
 
     def cp_script_to_remote
-      local_script_file = write_command_file_locally('script', sanitize_script(@script))
-      File.chmod(0555, local_script_file)
-      remote_script_file = remote_command_file('script')
-      upload_file(local_script_file, remote_script_file)
-      return remote_script_file
+      upload_data(sanitize_script(@script), remote_command_file('script'), 555)
     end
 
-    def upload_file(local_path, remote_path)
-      ensure_remote_directory(File.dirname(remote_path))
-      scp = Net::SCP.new(session)
-      upload_channel = scp.upload(local_path, remote_path)
-      upload_channel.wait
-    ensure
-      if upload_channel
-        upload_channel.close
-        upload_channel.wait
+    def upload_data(data, path, permissions = 555)
+      ensure_remote_directory File.dirname(path)
+      command = "#{su_prefix} tee '#{path}' >/dev/null && #{su_prefix} chmod '#{permissions}' '#{path}'"
```

Comment:
Running `sudo cat > $path` would fail if the actual user (not the sudo one) didn't have write permissions to `$path`, because the redirection is handled by the shell and not the process with elevated permissions